crates/ruff/src/rules/ruff/rules/unnecessary_list_allocation_for_first_element.rs

```diff
@@ -0,0 +1,186 @@
+            self.arg, self.arg
+        )
+    }
+
+    fn autofix_title(&self) -> String {
+        format!(
+            "Convert `list({})[0]` or equivalent list comprehension call to `next(iter({}))`",
+            self.arg, self.arg
+        )
+    }
+}
+
+/// RUF015
+pub(crate) fn unnecessary_list_allocation_for_first_element(
+    checker: &mut Checker,
+    call: &Expr,
+    slice: &Expr,
+    subscript_range: TextRange,
+) {
+    if !indexes_first_element(slice) {
+        return;
+    }
+    let Some(iter_name) = get_iterable_name(checker, call) else {
+        return;
+    };
+
+    let mut diagnostic = Diagnostic::new(
+        UnnecessaryListAllocationForFirstElement::new(iter_name.to_string()),
+        subscript_range,
+    );
+
+    if checker.patch(diagnostic.kind.rule()) {
+        let replacement = format!("next(iter({iter_name}))");
+        diagnostic.set_fix(Fix::suggested(Edit::range_replacement(
+            replacement,
+            subscript_range,
+        )));
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+/// Fetch the name of the iterable from a list expression if the expression returns an unmodified list
+/// which can be sliced into.
+fn get_iterable_name<'a>(checker: &mut Checker, expr: &'a Expr) -> Option<&'a str> {
+    match expr {
+        Expr::Call(ast::ExprCall { func, args, .. }) => {
+            let Some(id) = get_name_id(func.as_ref()) else {
+                return None;
+            };
+            if !(id == "list" && checker.semantic().is_builtin("list")) {
+                return None;
+            }
+
+            let Some(Expr::Name(ast::ExprName { id: arg_name, .. })) = args.first() else {
+                return None;
+            };
+
+            Some(arg_name.as_str())
+        }
+        Expr::ListComp(ast::ExprListComp {
+            elt, generators, ..
+        }) => {
+            // If the `elt` field is anything other than a [`Expr::Name`], we can't be sure that it
+            // doesn't modify the elements of the underlying iterator - for example, `[i + 1 for i in x][0]`.
+            if !matches!(elt.as_ref(), Expr::Name(ast::ExprName { .. })) {
+                return None;
+            }
+
+            // If there's more than 1 generator, we can't safely say that it fits the diagnostic conditions -
+            // for example, `[(i, j) for i in x for j in y][0]`.
+            if generators.len() != 1 {
+                return None;
+            }
+
+            let generator = &generators[0];
+            let Some(arg_name) = get_name_id(&generator.iter) else {
+                return None;
+            };
+            Some(arg_name)
+        }
+        _ => None,
+    }
+}
+
+/// Check that the slice [`Expr`] is functionally equivalent to slicing into the first element.
+fn indexes_first_element(expr: &Expr) -> bool {
+    match expr {
+        Expr::Constant(ast::ExprConstant { .. }) => {
+            get_effective_index(expr).unwrap_or(0i64) == 0i64
```

Comment:
Do you fall back to `0` here because any non int constants are not valid subscript indexes?

Reply:
Yep

Reply:
@MichaReiser should I do something else instead?